index.bs

```diff
@@ -5638,6 +5639,98 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+        ::  One or two inputs on which to evaluate PRF. Not all [=authenticators=] support evaluating the PRFs during credential creation so outputs may, or may not, be provided. If not, then an [=assertion=] is needed in order to obtain the outputs.
+
+        :   <dfn>evalByCredential</dfn>
+        ::  A record mapping [=base64url encoding|base64url encoded=] [=credential IDs=] to PRF inputs to evaluate for that credential. Only applicable during [=assertions=] when {{PublicKeyCredentialRequestOptions/allowCredentials}} is not empty.
+    </div>
+
+: Client extension processing ([=registration extension|registration=])
+::
+       1. If {{AuthenticationExtensionsPRFInputs/evalByCredential}} is present, return a {{DOMException}} whose name is “{{NotSupportedError}}”.
+       1. Set `hmac-secret` to [TRUE] in the authenticator extensions input.
+       1. If {{AuthenticationExtensionsPRFInputs/eval}} is present and a future extension to [[FIDO-CTAP]] permits evaluation of the PRF at creation time, configure `hmac-secret` inputs accordingly:
+           * Let `salt1` be the value of <code>SHA-256(UTF8Encode("WebAuthn PRF") || 0x00 || {{AuthenticationExtensionsPRFInputs/eval}}.{{AuthenticationExtensionsPRFValues/first}})</code>.
+           * If <code>{{AuthenticationExtensionsPRFInputs/eval}}.{{AuthenticationExtensionsPRFValues/second}}</code> is present, let `salt2` be the value of <code>SHA-256(UTF8Encode("WebAuthn PRF") || 0x00 || {{AuthenticationExtensionsPRFInputs/eval}}.{{AuthenticationExtensionsPRFValues/second}})</code>.
+       1. Set {{AuthenticationExtensionsPRFOutputs/enabled}} to the value of `hmac-secret` in the authenticator extensions output. If not present, set {{AuthenticationExtensionsPRFOutputs/enabled}} to [FALSE].
+       1. Set {{AuthenticationExtensionsPRFOutputs/results}} to the decrypted PRF result(s), if any.
+
+Note: If PRF results are obtained during [=registration=] then the [=[RP]=] MUST inspect the [=UV=] bit in the [=flags=] of the response in order to determine the correct value of {{PublicKeyCredentialRequestOptions/userVerification}} for future [=assertions=]. Otherwise results from [=assertions=] may be inconsistent with those from the [=registration=].
+
+: Client extension processing ([=authentication extension|authentication=])
+::
+       1. If either {{AuthenticationExtensionsPRFInputs/eval}} is present or {{AuthenticationExtensionsPRFInputs/evalByCredential}} is not empty:
+           1. If {{PublicKeyCredentialRequestOptions/userVerification}} has the value {{UserVerificationRequirement/preferred}}, return a {{DOMException}} whose name is “{{NotSupportedError}}”.
+           1. If {{AuthenticationExtensionsPRFInputs/evalByCredential}} is not empty but {{PublicKeyCredentialRequestOptions/allowCredentials}} is empty, return a {{DOMException}} whose name is “{{NotSupportedError}}”.
+           1. Initialize the {{AuthenticationExtensionsClientOutputs/prf}} extension output to an empty dictionary.
```

Comment:
Hm, this step was meant to always run, wasn't it? In that case it would have to move to before the parent step. But on the other hand, I think that check ("If either eval is present or evalByCredential is not empty") is now also unnecessary? With the new logic below I think that if `eval` is absent and `evalByCredential` is empty, then _ev_ will just end up null and the output will remain empty (but present).

Reply:
This was deliberate: echoing an extension at assertion time means that the user agent processed the extension, tried, but wasn't able to get a response. (Maybe the authenticator doesn't support hmac_secret.)

But if the browser didn't find any inputs then that's not true. I.e. it's an RP error, and echoing the extension like everything is ok would be misleading.

Reply:
Ok, but in that case, shouldn't some "invalid input" error be raised instead? [§9.2. Defining Extensions](https://w3c.github.io/webauthn/#sctn-extension-specification) reads:

>Any client extension that is processed by the client MUST return a client extension output value so that the WebAuthn Relying Party knows that the extension was honored by the client.

So if the extension output is missing, that generally means the extension was ignored or not supported by either the browser or authenticator, which would not be the case here. The `appidExclude` extension goes against this, but we should probably change that since the spec is not internally consistent as is...

Reply:
> Ok, but in that case, shouldn't some "invalid input" error be raised instead?

That would be fine too, but it would preclude adding anything else in this extension in the future unless it always comes with inputs in the existing members. I'm ok with that, I just didn't see a reason to do it.

> So if the extension output is missing, that generally means the extension was ignored or not supported by either the browser or authenticator, which would not be the case here.

But that _is_ the case here. The RP provided no inputs so the extension was ignored.

> The appidExclude extension goes against this, but we should probably change that

appidExclude shouldn't have an output because there's nothing the RP can do with that information. I'm sure that paragraph in 9.2 made rough sense at the time, but it hasn't proven to be true and I think we should delete it.

Reply:
>But that is the case here. The RP provided no inputs so the extension was ignored.

Ok, I see what you mean. Fair enough.

Let's track the appidExclude thing in #1445.